pkg/strategy/infinity-grid/strategy.go

```diff
@@ -0,0 +1,534 @@
+
+	LowerPrice fixedpoint.Value `json:"lowerPrice" yaml:"lowerPrice"`
+
+	// Buy-Sell Margin for each pair of orders
+	Margin fixedpoint.Value `json:"margin"`
+
+	// Quantity is the quantity you want to submit for each order.
+	Quantity fixedpoint.Value `json:"quantity"`
+
+	InitialOrderQuantity fixedpoint.Value `json:"initialOrderQuantity"`
+	CountOfMoreOrders    int              `json:"countOfMoreOrders"`
+
+	// GridNum is the grid number, how many orders you want to post on the orderbook.
+	GridNum int64 `json:"gridNumber" yaml:"gridNumber"`
+
+	// Side is the initial maker orders side. defaults to "both"
+	Side types.SideType `json:"side" yaml:"side"`
+
+	// Long means you want to hold more base asset than the quote asset.
+	Long bool `json:"long,omitempty" yaml:"long,omitempty"`
+
+	state *State
+
+	// orderStore is used to store all the created orders, so that we can filter the trades.
+	orderStore *bbgo.OrderStore
+
+	// activeOrders is the locally maintained active order book of the maker orders.
+	activeOrders *bbgo.ActiveOrderBook
+
+	tradeCollector *bbgo.TradeCollector
+
+	currentUpperGrid int
+	currentLowerGrid int
+
+	// groupID is the group ID used for the strategy instance for canceling orders
+	groupID uint32
+}
+
+func (s *Strategy) ID() string {
+	return ID
+}
+
+func (s *Strategy) Validate() error {
+	if s.LowerPrice.IsZero() {
+		return errors.New("lowerPrice can not be zero, you forgot to set?")
+	}
+
+	if s.Margin.Sign() <= 0 {
+		// If margin is empty or its value is negative
+		return fmt.Errorf("Margin should bigger than 0")
+	}
+
+	if s.Quantity.IsZero() {
+		return fmt.Errorf("Quantity can not be zero")
+	}
+
+	return nil
+}
+
+func (s *Strategy) placeInfiniteGridOrders(orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) {
+	balances := session.Account.Balances()
+	log.Infof("Balances: %s", balances.String())
+	log.Infof("Base currency: %s", s.Market.BaseCurrency)   // BTC
+	log.Infof("Quote currency: %s", s.Market.QuoteCurrency) // USD
+	baseBalance, ok := balances[s.Market.BaseCurrency]
+	if !ok {
+		log.Errorf("base balance %s not found", s.Market.BaseCurrency)
+		return
+	}
+	if s.currentUpperGrid != 0 || s.currentLowerGrid != 0 {
+		// reconnect, do not place orders
+		return
+	}
+
+	quoteBalance, ok := balances[s.Market.QuoteCurrency]
+	if !ok || quoteBalance.Available <= 0 { // check available USD in balance
+		log.Errorf("quote balance %s not found", s.Market.QuoteCurrency)
+		return
+	}
+
+	var orders []types.SubmitOrder
+	var quantityF fixedpoint.Value
+	currentPrice, ok := session.LastPrice(s.Symbol)
+	if !ok {
+		return
+	}
+
+	if false {
+		if s.InitialOrderQuantity > 0 {
+			quantityF = s.InitialOrderQuantity
+		} else {
+			quantityF = s.Quantity.Div(fixedpoint.NewFromFloat(1.0 - 1/(1.0+s.Margin.Float64())))
+		}
+
+		// Buy half of value of asset
+		order := types.SubmitOrder{
+			Symbol:      s.Symbol,
+			Side:        types.SideTypeBuy,
+			Type:        types.OrderTypeLimit,
```

Comment:
@c9s if I set market order, backtest fail with error: 
<img width="1585" alt="image" src="https://user-images.githubusercontent.com/8588439/174246836-e809e0db-3d5b-427e-a9fc-698cf657c660.png">

can you help me to fix it?

Reply:
maybe you forgot to sync first?

Reply:
yes, you're missing the kline data, you need to sync first.

Reply:
I tried, but have the same error: 
`go run ./cmd/bbgo backtest -v --sync --config config/infinity-grid.yaml --base-asset-baseline --output output --subdir`

Reply:
The reason is that you submit an order before the first kline is processed. 

![image](https://user-images.githubusercontent.com/50894/174388981-4b54acb7-19bc-4ba8-9ff4-9e99f10e34c1.png)

Reply:
check if your db loaded all the binance klines, also you might need to rebase to the latest main branch.

Reply:
I already rebase my branch, make some fixes, but error not left. 

Can you pls debug locally? just run 
> go run ./cmd/bbgo backtest -v --sync --config config/infinity-grid.yaml --base-asset-baseline --output output --subdir

I think it's bug, couse I have last price https://github.com/c9s/bbgo/pull/735/files#diff-378c135c835046a31b907fc9e387f6d6a287d76248e6261d87689b4147772108R138 and use it in this stage, but I can evaluates market order in stream onStart stage.

Reply:
If the post order action happens in OnStart, you should get the error "SubmitOrders should be called after UserDataStream been initialized"